tests/DCPS/Compiler/xcdr/xcdr.cpp

```diff
@@ -0,0 +1,571 @@
+{
+  expect_values_equal(a, b);
+  if (::testing::Test::HasFailure()) {
+    return ::testing::AssertionFailure() << a_expr << " != " << b_expr;
+  }
+  return ::testing::AssertionSuccess();
+}
+
+struct DataView {
+  DataView(const std::string& s)
+  : data(s.data())
+  , size(s.size())
+  {
+  }
+
+  DataView(const ACE_Message_Block& mb)
+  : data(mb.base())
+  , size(mb.length())
+  {
+  }
+
+  const char* data;
+  size_t size;
+};
+
+bool operator==(const DataView& a, const DataView& b)
+{
+  return a.size == b.size ? !std::memcmp(a.data, b.data, a.size) : false;
+}
+
+::testing::AssertionResult assert_DataView(
+  const char* a_expr, const char* b_expr,
+  const DataView& a, const DataView& b)
+{
+  if (a == b) {
+    return ::testing::AssertionSuccess();
+  }
+
+  std::string a_line;
+  std::stringstream astrm(to_hex_dds_string(a.data, a.size, '\n', 8));
+  std::string b_line;
+  std::stringstream bstrm(to_hex_dds_string(b.data, b.size, '\n', 8));
+  std::stringstream result;
+  bool got_a = true;
+  bool got_b = true;
+  bool diff = true;
+  while (got_a || got_b) {
+    result.width(16);
+    result << std::left;
+    if (got_a && std::getline(astrm, a_line, '\n')) {
+      result << a_line;
+    } else {
+      got_a = false;
+      result << ' ';
+    }
+    if (got_b && std::getline(bstrm, b_line, '\n')) {
+      result << " " << b_line;
+    } else {
+      got_b = false;
+    }
+    if (diff && a_line != b_line) {
+      result << " <- Different Starting Here";
+      diff = false;
+    }
+    result << std::endl;
+  }
+  return ::testing::AssertionFailure()
+    << a_expr << " (size " << a.size << ") isn't the same as "
+    << b_expr << " (size " << b.size << ").\n"
+    << a_expr << " is on the left, " << b_expr << " is on the right:\n"
+    << result.str();
+}
+
+template<typename TypeA, typename TypeB>
+void amalgam_serializer_test(
+  const Encoding& encoding, const std::string expected_cdr, TypeA& value, TypeB& result)
+{
+  ACE_Message_Block buffer(1024);
+
+  // Serialize and Compare CDR
+  {
+    Serializer serializer(&buffer, encoding);
+    EXPECT_TRUE(serializer << value);
+    EXPECT_PRED_FORMAT2(assert_DataView, expected_cdr, buffer);
+  }
+
+  // Deserialize and Compare C++ Values
+  {
+    Serializer serializer(&buffer, encoding);
+    EXPECT_TRUE(serializer >> result);
+    EXPECT_PRED_FORMAT2(assert_values, value, result);
+  }
+}
+
+template<typename TypeA, typename TypeB>
+void amalgam_serializer_test(const Encoding& encoding, const std::string expected_cdr)
+{
+  TypeA value;
+  set_values(value);
+  TypeB result;
+  amalgam_serializer_test<TypeA, TypeB>(encoding, expected_cdr, value, result);
+}
+
+template<typename Type>
+void serializer_test(const Encoding& encoding, const std::string expected_cdr)
+{
+  amalgam_serializer_test<Type, Type>(encoding, expected_cdr);
+}
+
+// XCDR1 =====================================================================
+
+const std::string xcdr1_non_mutable_expected(
```

Comment:
It would be easier if the compiler did the size. I think my reasoning here was what it can be made using an anonymous object to pass to functions where I don't think you can do that with an array, at least in 03.

Reply:
It would still be a named object, just like `xcdr1_non_mutable_expected` is.
https://godbolt.org/z/NidKrs